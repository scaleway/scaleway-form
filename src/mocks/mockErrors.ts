import type { FormErrors } from '../types'

export const emailRegex = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i

export const mockErrors: FormErrors = {
  MAX_DATE: ({ maxDate }) => `Date must be lower than ${maxDate?.toString()}`,
  MAX_LENGTH: ({ maxLength }) =>
    `This field should have a length lower than ${maxLength}`,
  MIN_DATE: ({ minDate }) => `Date must be greater than ${minDate?.toString()}`,
  MIN_LENGTH: ({ minLength }) =>
    `This field should have a length greater than ${minLength}`,
  REGEX: ({ regex }) => {
    if (
      (Array.isArray(regex[0])
        ? regex?.[0]?.[0]?.source
        : regex?.[0]?.source) === emailRegex.source
    ) {
      return 'Invalid email'
    }

    return `This field should match the regex ${regex
      .map(r =>
        Array.isArray(r)
          ? r.map(nestedRegex => nestedRegex.source).join(' or ')
          : r.source,
      )
      .join(' and ')}`
  },
  REQUIRED: 'This field is required',
  TOO_HIGH: ({ max }) => `This field is too high (maximum is : ${max})`,
  TOO_LOW: 'This field is too low',
}
