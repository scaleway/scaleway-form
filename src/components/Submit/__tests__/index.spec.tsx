import userEvent from '@testing-library/user-event'
import { Form, Submit, TextBoxField } from '../..'
import {
  mockRandom,
  restoreRandom,
  shouldMatchEmotionSnapshot,
  shouldMatchEmotionSnapshotFormWrapper,
} from '../../../../.jest/helpers'
import { mockErrors } from '../../../mocks'

const alpha = /^[a-zA-Z]*$/

describe('Submit', () => {
  test('renders correctly ', () =>
    shouldMatchEmotionSnapshotFormWrapper(<Submit>Test</Submit>))

  test('renders correctly with icon and iconPosition ', () =>
    shouldMatchEmotionSnapshotFormWrapper(
      <Submit icon="east" iconPosition="right">
        Test
      </Submit>,
    ))

  test('form is invalid', () =>
    shouldMatchEmotionSnapshot(
      <Form
        onRawSubmit={() => {}}
        initialValues={{ toto: '4' }}
        errors={mockErrors}
      >
        <TextBoxField name="toto" regex={[alpha]} />
        <Submit>Test</Submit>
      </Form>,
    ))

  test('form is submitting', async () => {
    mockRandom()

    await shouldMatchEmotionSnapshot(
      <Form
        onRawSubmit={() =>
          new Promise(resolve => {
            setTimeout(() => resolve(undefined), 5000)
          })
        }
        errors={mockErrors}
      >
        <Submit>Test</Submit>
      </Form>,
      {
        transform: async ({ getByText }) => {
          await userEvent.click(
            getByText('Test').closest('button') as HTMLButtonElement,
          )
          expect(
            getByText('Test').closest('button') as HTMLButtonElement,
          ).toBeDisabled()
        },
      },
    )

    restoreRandom()
  })
})
