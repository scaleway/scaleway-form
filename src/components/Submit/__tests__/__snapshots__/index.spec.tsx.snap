// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit form is invalid 1`] = `
<DocumentFragment>
  .cache-bjn8wh {
  position: relative;
}

.cache-1l46zoq {
  -webkit-transition: border-color 0.2s ease,box-shadow 0.2s ease;
  transition: border-color 0.2s ease,box-shadow 0.2s ease;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: #ffffff;
  background-image: none;
  border: 1px solid #d4dae7;
  border-radius: 4px;
  color: #4a4f62;
  display: block;
  max-width: 100%;
  outline: none;
  position: relative;
  width: 100%;
  padding-left: 8px;
  padding-right: 8px;
  padding-top: 14px;
  font-size: 16px;
  line-height: 24px;
  height: 48px;
  padding-left: 8px;
  padding-right: 20px;
  padding-top: 14px;
  padding: 8px;
  border-color: #dd3252;
}

.cache-1l46zoq::-webkit-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq::-moz-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq:-ms-input-placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq::placeholder {
  color: #b2b6c3;
  opacity: 0;
}

.cache-1l46zoq:hover,
.cache-1l46zoq:focus {
  border-color: #4f0599;
}

.cache-1l46zoq:focus {
  box-shadow: 0px 0px 0px 3px #4F059940;
  border-color: #4f0599;
}

.cache-1l46zoq::-webkit-input-placeholder {
  opacity: 1;
}

.cache-1l46zoq::-moz-placeholder {
  opacity: 1;
}

.cache-1l46zoq:-ms-input-placeholder {
  opacity: 1;
}

.cache-1l46zoq::placeholder {
  opacity: 1;
}

.cache-1l46zoq:hover,
.cache-1l46zoq:focus {
  border-color: #a6102d;
}

.cache-1l46zoq:focus {
  box-shadow: 0px 0px 0px 3px #DD325240;
  border-color: #a6102d;
}

.cache-msb198 {
  -webkit-transition: max-height 300ms ease-out,opacity 300ms ease-out;
  transition: max-height 300ms ease-out,opacity 300ms ease-out;
  max-height: 0;
  overflow: hidden;
  opacity: 0;
  height: auto;
  margin-top: 0;
  -webkit-transition: max-height 300ms ease-in,opacity 300ms ease-in;
  transition: max-height 300ms ease-in,opacity 300ms ease-in;
  max-height: 56px;
  opacity: 1;
  overflow: visible;
  overflow: hidden;
}

.cache-1rshnsu {
  font-size: 12px;
  color: #dd3252;
  padding-top: 2px;
}

.cache-1f4ynuo {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border-width: 0;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  white-space: nowrap;
  font-weight: 500;
  -webkit-transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  font-size: 16px;
  line-height: 32px;
  font-weight: 500;
  padding: 8px 16px;
  background-color: #22c8a2;
  color: #ffffff;
  cursor: default;
  pointer-events: none;
  color: #c4c9d6;
  background-color: #f6f6f8;
  border-color: #d4dae7;
  box-shadow: none;
}

.cache-1f4ynuo:hover,
.cache-1f4ynuo:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.cache-1f4ynuo:hover,
.cache-1f4ynuo:focus {
  color: #ffffff;
  background-color: #149a7b;
}

.cache-1f4ynuo:focus {
  box-shadow: 0 0 0 2px #149a7b40;
}

.cache-1f4ynuo svg {
  fill: #c4c9d6;
}

.cache-115wth8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  pointer-events: none;
}

<form
    novalidate=""
  >
    <div>
      <div
        class="cache-bjn8wh e18bmnpq5"
      >
        <input
          autocomplete="on"
          class="cache-1l46zoq e18bmnpq1"
          name="toto"
          type="text"
          value="4"
        />
      </div>
      <div
        class="cache-msb198 e18bmnpq2"
      >
        <div
          class="cache-1rshnsu e18bmnpq4"
        >
          This field should match the regex ^[a-zA-Z]*$
        </div>
      </div>
    </div>
    <button
      aria-disabled="true"
      class="cache-1f4ynuo em6gco80"
      disabled=""
      type="submit"
    >
      <div
        class="cache-115wth8 em6gco83"
      >
        Test
      </div>
    </button>
  </form>
</DocumentFragment>
`;

exports[`Submit form is submitting 1`] = `
<DocumentFragment>
  @keyframes animation-0 {
  from {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  to {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

.cache-1f4ynuo {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border-width: 0;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  white-space: nowrap;
  font-weight: 500;
  -webkit-transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  font-size: 16px;
  line-height: 32px;
  font-weight: 500;
  padding: 8px 16px;
  background-color: #22c8a2;
  color: #ffffff;
  cursor: default;
  pointer-events: none;
  color: #c4c9d6;
  background-color: #f6f6f8;
  border-color: #d4dae7;
  box-shadow: none;
}

.cache-1f4ynuo:hover,
.cache-1f4ynuo:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.cache-1f4ynuo:hover,
.cache-1f4ynuo:focus {
  color: #ffffff;
  background-color: #149a7b;
}

.cache-1f4ynuo:focus {
  box-shadow: 0 0 0 2px #149a7b40;
}

.cache-1f4ynuo svg {
  fill: #c4c9d6;
}

.cache-i7e9al {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 8px;
  pointer-events: none;
}

.cache-1dul7yn-css {
  -webkit-animation: animation-0 0.75s linear infinite;
  animation: animation-0 0.75s linear infinite;
}

.cache-11f05l0-Loader {
  -webkit-transition: stroke-dashoffset 0.5s ease 0s;
  transition: stroke-dashoffset 0.5s ease 0s;
}

.cache-115wth8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  pointer-events: none;
}

<form
    novalidate=""
  >
    <button
      aria-disabled="true"
      class="cache-1f4ynuo em6gco80"
      disabled=""
      type="submit"
    >
      <div
        class="cache-i7e9al em6gco81"
      >
        <svg
          aria-label="Loading"
          aria-valuemax="100"
          aria-valuemin="0"
          aria-valuenow="20"
          aria-valuetext="20%"
          class="cache-1dul7yn-css"
          role="progressbar"
          style="height: 1em; width: 1em;"
          viewBox="0 0 100 100"
        >
          <circle
            cx="50"
            cy="50"
            fill="none"
            r="42"
            stroke="transparent"
            stroke-width="16"
          />
          <circle
            class="cache-11f05l0-Loader"
            cx="50"
            cy="50"
            fill="none"
            r="42"
            stroke="currentColor"
            stroke-dasharray="263.89378290154264"
            stroke-dashoffset="211.11502632123413"
            stroke-linecap="round"
            stroke-width="16"
          />
        </svg>
      </div>
      <div
        class="cache-115wth8 em6gco83"
      >
        Test
      </div>
    </button>
  </form>
</DocumentFragment>
`;

exports[`Submit renders correctly  1`] = `
<DocumentFragment>
  .cache-107ul6f {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border-width: 0;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  white-space: nowrap;
  font-weight: 500;
  -webkit-transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  font-size: 16px;
  line-height: 32px;
  font-weight: 500;
  padding: 8px 16px;
  background-color: #22c8a2;
  color: #ffffff;
}

.cache-107ul6f:hover,
.cache-107ul6f:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.cache-107ul6f:hover,
.cache-107ul6f:focus {
  color: #ffffff;
  background-color: #149a7b;
}

.cache-107ul6f:focus {
  box-shadow: 0 0 0 2px #149a7b40;
}

.cache-115wth8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  pointer-events: none;
}

<form
    novalidate=""
  >
    <button
      aria-disabled="false"
      class="cache-107ul6f em6gco80"
      type="submit"
    >
      <div
        class="cache-115wth8 em6gco83"
      >
        Test
      </div>
    </button>
  </form>
</DocumentFragment>
`;

exports[`Submit renders correctly with icon and iconPosition  1`] = `
<DocumentFragment>
  .cache-107ul6f {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border-width: 0;
  cursor: pointer;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  outline: none;
  vertical-align: middle;
  white-space: nowrap;
  font-weight: 500;
  -webkit-transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  transition: color 150ms ease-in-out,background-color 150ms ease-in-out,border-color 150ms ease-in-out;
  font-size: 16px;
  line-height: 32px;
  font-weight: 500;
  padding: 8px 16px;
  background-color: #22c8a2;
  color: #ffffff;
}

.cache-107ul6f:hover,
.cache-107ul6f:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.cache-107ul6f:hover,
.cache-107ul6f:focus {
  color: #ffffff;
  background-color: #149a7b;
}

.cache-107ul6f:focus {
  box-shadow: 0 0 0 2px #149a7b40;
}

.cache-115wth8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  pointer-events: none;
}

.cache-14frhxz {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 8px;
  pointer-events: none;
}

.cache-kv8vgz-sizeStyles {
  vertical-align: middle;
  fill: currentColor;
  height: 1em;
  width: 1em;
  min-width: 1em;
  min-height: 1em;
}

<form
    novalidate=""
  >
    <button
      aria-disabled="false"
      class="cache-107ul6f em6gco80"
      type="submit"
    >
      <div
        class="cache-115wth8 em6gco83"
      >
        Test
      </div>
      <div
        class="cache-14frhxz em6gco81"
      >
        <svg
          class="cache-kv8vgz-sizeStyles etwatq50"
          viewBox="0 0 24 24"
        >
          <path
            d="M15 5l-1.41 1.41L18.17 11H2v2h16.17l-4.59 4.59L15 19l7-7-7-7z"
          />
        </svg>
      </div>
    </button>
  </form>
</DocumentFragment>
`;
